### ISSUE TITLE
Suggestion: `compactMap` operator

### ISSUE THREAD

**Issue by takasek** [ReactiveX/RxSwift#1764]
We use `flatMap` to transform and filter events.

```swift
stream.flatMap { x -> Observable<Foo> in
    do {
        try x.validate()
        guard x.isOnCertainCondition else { return .empty() }
        return x.fooOrNil().map(Observable.just) ?? .empty()
    } catch let e {
        return .error(e)
    }
}

optionalStream
    .flatMap(Observable.from(optional:))
```

But flatMapping seems too much as a solution in some cases.
What do you think to introduce `compactMap` to RxSwift?
With `func compactMap<U>(transform: @escaping (E) throws -> U?) -> Observable<U>` we can replace the former code like below.

```swift
stream.compactMap { x -> Foo? in
    try x.validate()
    guard x.isOnCertainCondition else { return nil }
    return x.fooOrNil()
}

optionalStream
    .compactMap { $0 }
```

This solution seems more natural for me.
It is similar to [FilterMap](https://github.com/RxSwiftCommunity/RxSwiftExt/blob/2204a685f7761244f311161f6c35ea0406488e17/Source/RxSwift/filterMap.swift) but `compactMap` provides more swifty way.

However, I saw a discussion on `Rxswift's flatMap operator works differently as it is in Swift` #1712
@freak4pc said:

> There is no correlation between Swift's compactMap and flatMap situation to what Rx's flatMap aims to solve, since they work on entirely different data types and solve very different things.

Is it negative to conform RxSwift to Swift style?

**Comment by freak4pc** [comment]
Hey @takasek - Thanks for taking the time to draft this :-)

I'm personally -1 on this suggestion for several reason: 

1. We should strive to only add useful operators that are *part of the ReactiveX.io* standard. Adding too much code (especially relatively simplistic features) just bloats the codebase, thus also the maintenance and supports the open source team needs to provides.

2. In the issue you linked two you can see the second issue which is even worst for me. The `flatMap` term is shared between what we know from Swift and `flatMap` the Rx operator. This is already confusing for some people. Adding `compactMap` into the fire would further increase that confusion in my honest opinion. 

tl;dr - I would think more on this if this operator would provide extremely significant value, but it is quite trivial to create in your own codebase if you need it. 

> Is it negative to conform RxSwift to Swift style?

I believe its not only negative, but impossible. As I mentioned, there is a standard, and we should strive to keep a unified standard with ReactiveX to make sure the platform-agnostic terminology is persisted.  Otherwise, teams working with RxJava and RxSwift, for example, have different meaning for what operators do, which is extremely counter-productive and against what Rx tries to achieve. 

That's my opinion of it :)

**Comment by takasek** [comment]
Hi @freak4pc , thank you for replying!

Your opinion makes sense.
RxSwift should follow ReactiveX standard.  And we strive to keep the codebase minimum.

I just reached to one of point of a discussion.
The point is whether `compactMap` is a general idea for ReactiveX's Observable sequence, or locally extended idea in `Swift.Sequence`.

[ReactiveX's `Observable` is a sequence](https://github.com/ReactiveX/RxSwift/blob/master/Documentation/GettingStarted.md#basics).
Both `Observable` and `Swift.Sequence` are abstractions of general sequence (monad), and both they provide general functionalities to handle their elements iteratively.
From this point of view, `Observable.flatMap` shares entirely same functionality with `Swift.Sequence.flatMap`.

So, how is `Swift.Sequence.compactMap`?
If we can apply the idea of `compactMap` across other platforms' sequence, introducing `compactMap` into `Observable` may be reasonable.

Referring to [a proposal examined to introduce `compactMap` to Swift standard library](https://github.com/apple/swift-evolution/blob/master/proposals/0187-introduce-filtermap.md#proposed-solution), you can see same functionality in other languages:

> - Haskell, Idris
  `mapMaybe :: (a -> Maybe b) -> [a] -> [b]`
> - Ocaml (Core and Batteries)
  `filter_map : 'a t -> f:('a -> 'b option) -> 'b t`
> - F#
  `List.choose : ('T -> 'U option) -> 'T list -> 'U list`
> - Rust
  `fn filter_map<B, F>(self, f: F) -> FilterMap<Self, F> where F: FnMut(Self::Item) -> Option<B>`
> - Scala
  `def collect[B](pf: PartialFunction[A, B]): List[B]`

It may indicate that `compactMap` is a general and useful idea for sequence across platforms.

Please note that I don't have so strong motivation to force introducing `compactMap` to RxSwift core.
Even if `compactMap` is reasonable, we need to consider these tradeoffs:

- Bloating codebase affects maintenance
- In addition, new idea costs users to learn

Suggesting this draft at `RxSwiftExt` may be better.
Also, as you mention, it is not difficult to implement `compactMap` in my own codebase.

But I currently guess that `compactMap` may penetrate other ReactiveX's `Observable` over RxSwift.

**Comment by kzaher** [comment]
So,

I guess we try to make it consistent with other Rx implementations, although every implementation does have its own set off adaptations for a particular environment.

I guess I had used `compactMap` on an `Observable` sequence once or twice. It's unclear to me is there a general need for it.

When this project started almost 4 years ago we didn't really think about aligning with Swift guidelines because I don't there were any at that time. Compiler was crashing or generating malformed binary on a daily basis, we were using `>-` because using extensions wasn't possible.

Right now it looks like Swift is maturing. I think that `compactMap` was originally called `flatMap`. I think it was renamed into `compactMap` recently.

The summary of all of this would probably be:
* In the future we will also try to make the library more consistent with `Swift` APIs. I guess there are some places where we can improve it.
* I would still like to give Swift standard library a bit more time to prove it has stabilized. If the next major release (Swift 5.0) will contain `compactMap` we can probably assume that this is a stable API and revise this issue.
* For now I would suggest waiting a bit. This isn't a hard to implement API, so adding it to `SwiftExt` for now would make more sense.

I would really like to understand what is the statistic of operator usage in projects. Maybe there are some operators we can safely remove and replace them with others ones.

One way could be making a tool that crawls projects and uses SourceKit to provide the statistics. Have no idea how to solve trustworthiness problem and reject manipulations. 

Having the actual data on the subject might help us to make better decisions in the future. How to solve this issue in general way, have no idea for now.

**Comment by hmlongco** [comment]
kzaher,

Just had a use for this yesterday when I attempted to use the new Swift 5 Result type...

```
class myViewModel {

    var data: Observable<AccountInformation>!
    var processing: Observable<Bool>!
    var error: Observable<String>!

    func actions(_ load: Observable<Void>) {
        let loading = load
            .flatMapLatest { [unowned self] (_) -> Observable<Result<AccountInformation, Error>> in
                return self.accountInfoAPI()
                    .map { .success($0) }
                    .catchError { error in .just(.failure(error)) }
            }
            .share()

        data = loading
            .compactMap { $0.value() }

        error = loading
            .compactMap { $0.error()?.localizedDescription }

        processing = Observable
            .merge(loading.map{_ in true}, data.map{_ in false}, error.map{_ in false})
    }

}
```
Where compactMap was...
```
    func compactMap<R>(_ transform: @escaping (Self.E) throws -> R?) -> RxSwift.Observable<R> {
        return self.map { try? transform($0) }.filter { $0 != nil }.map { $0! }
    }
```
And actually....
```
public extension Swift.Result where Failure : Error {
    func value() -> Success? {
        switch self {
        case .success(let success):
            return success
        case .failure(_):
            return nil
        }
    }
    func error() -> Failure? {
        switch self {
        case .success(_):
            return nil
        case .failure(let error):
            return error
        }
    }
}
```
As Apple had try? result.get() but no corresponding operator to fetch the failure value.

Bottom line though is that optionals are part and parcel of RxSwift and RxSwift could be somewhat better at dealing with them. (Also the oft-mentioned 'unwrap' operator.

**Comment by kzaher** [comment]
@hmlongco 
```
let loading = load
            .flatMapLatest { [unowned self] (_) -> Observable<Result<AccountInformation, Error>> in
                return self.accountInfoAPI()
                    .map { .success($0) }
                    .catchError { error in .just(.failure(error)) }
            }
            .share()

        data = loading
            .compactMap { try? $0.get() }

        error = loading
            .compactMap { try? $0.error().localizedDescription }

        processing = Observable
            .merge(loading.map{_ in true}, data.map{_ in false}, error.map{_ in false})
```

Please don't write code like this :)

We can add `compactMap`, but please don't use it in this way.

**Comment by hmlongco** [comment]
@kzaher 

So I guess I'm missing something. I have a trigger action that flapmaps the API call and the result feeds three different observables: one drives a view (I need the view), one drives a processing spinner (I need to show activity in progress), and one drives an error message (if any, but I still need to show it if it occurs).

How else would you do it, assuming you want to avoid Subjects in the VM?

**Comment by hmlongco** [comment]
@kzaher So if I wanted to do a PR adding compactMap, would we do it simply as...
```
extension ObservableType {
    func compactMap<R>(_ transform: @escaping (Self.E) throws -> R?) -> RxSwift.Observable<R> {
        return self.map { try? transform($0) }.filter { $0 != nil }.map { $0! }
    }
}
```
Or would we want to add it as a full-fledged operator with its own class, sink, etc.. so as to avoid  generating three operations, sinks, disposables, etc.. behind the scenes each time it was used?

**Comment by kzaher** [comment]
@hmlongco Like map, not a composition.

**Comment by hmlongco** [comment]
Somewhat cleaner example using the materialize operator instead of Result.
```
    func actions(_ load: Observable<Void>) {
        let loading = load
            .flatMapLatest { AccountInformation.load().materialize() }
            .share()

        accountInformation = loading
            .compactMap { $0.element }
            .share()

        error = loading
            .compactMap { $0.error?.localizedDescription }

        processing = Observable
            .merge(load.map{_ in true}, loading.map{_ in false})
    }
```
